File: py/haltrule/checkpoint.py
# Ported from ts/checkpoint.ts. Pure policy: no I/O, no timers, no clock.
# fixtures/checkpoint/v0.json is the contract both language ports must satisfy -
# do not change behavior here without checking it still passes that fixture.
# Behavior (not names) must match the TypeScript original; Python identifiers
# follow snake_case instead of the source's camelCase.
#
# Where JavaScript and Python disagree about a value, this file follows
# JavaScript, because the spec's value model was defined by what JavaScript
# can represent: 1.0 is the integer 1, map keys sort by UTF-16 code unit (not
# code point), and "is this missing?" - asked of what an artifact records -
# uses JavaScript falsiness, under which an empty list or dict is present.
#
# The arguments are typed: an argument of another type is refused (TypeError),
# and None or not given is the one way to be absent. A key the signature does
# not name is refused by the call itself.

from __future__ import annotations

import hashlib
import math
from typing import Any, Literal, Mapping, Optional, Union

from haltrule.messages import describe_number
from haltrule.verdict import VerdictLevel, is_verdict_level, verdict

# --------------------------------------------------------------- canonicalize

# Largest integer every implementation represents exactly: 2^53 - 1.
_MAX_SAFE_INTEGER = 9007199254740991

# Deepest nesting of lists and maps a digest input may have. A bound every
# language can reach without exhausting its stack, so all of them halt at the
# same depth instead of each crashing at its own.
_MAX_DEPTH = 100

DigestInputReason = Literal[
    "digest_input_float",
    "digest_input_int_range",
    "digest_input_unsupported",
]


class _DigestInputError(ValueError):
    """Internal: unwinds the recursive encoder to the public boundary, where
    it becomes a halt verdict. Its message names where the value sits
    (`$.a[2]`), a path being no thing two languages spell alike."""

    def __init__(self, reason: DigestInputReason, message: str) -> None:
        super().__init__(message)
        self.reason = reason


# The seven characters with a two-character escape. Every other code point
# below U+0020 is written \u00xx (lowercase hex); everything else, "/",
# U+007F, U+2028 and U+2029 included, is written as itself - the same output
# as the TypeScript port's table and as json.dumps(..., ensure_ascii=False).
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
    0x22: '\\"',
    0x5C: "\\\\",
}


def canonicalize(value: Any) -> dict[str, str]:
    """Canonical form of a value: an RFC 8785 subset.

    Map keys are sorted by UTF-16 code unit, there is no insignificant
    whitespace, integers are written in decimal, and strings use the escape
    table above. Strings are not Unicode-normalized: NFC and NFD spellings of
    the same text canonicalize differently.

    Accepted: None, booleans, integral numbers with |n| <= 2^53 - 1 (so 1.0 is
    the integer 1, as it is in JavaScript), strings of Unicode scalar values,
    lists, and dicts with str keys, nested at most 100 deep.

    Returns {"canonical": ...}, or a halt verdict for anything else. Like
    every verdict here it is a value, never a raise: the caller records it
    before it stops.
    """
    try:
        return {"canonical": _encode_value(value, "$", 0)}
    except _DigestInputError as error:
        return verdict("halt", error.reason, str(error))


def checkpoint_digest(value: Any) -> dict[str, str]:
    """Return {"digest": "sha256:" + the lowercase hex SHA-256 of the canonical
    form's UTF-8 bytes} - the value `printf '%s' "$canonical" | sha256sum`
    prints - or the halt canonicalize returned."""
    result = canonicalize(value)
    if "verdict" in result:
        return result
    return {
        "digest": "sha256:"
        + hashlib.sha256(result["canonical"].encode("utf-8")).hexdigest()
    }


def _utf16_key(key: str) -> bytes:
    """Sort key ordering strings by UTF-16 code unit, as JavaScript does.

    UTF-16-BE bytes compare exactly as code units do; surrogatepass lets an
    unpaired surrogate through to _encode_string, which rejects it.
    """
    return key.encode("utf-16-be", "surrogatepass")


def _encode_value(value: Any, at: str, depth: int) -> str:
    if value is None:
        return "null"
    # bool before int: bool is an int subclass, and True must not become 1.
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return _encode_number(value, at)
    if isinstance(value, str):
        return _encode_string(value, at)
    if isinstance(value, (list, dict)) and depth >= _MAX_DEPTH:
        raise _DigestInputError(
            "digest_input_unsupported",
            f"{at}: nested deeper than {_MAX_DEPTH} lists and maps",
        )
    if isinstance(value, list):
        return (
            "["
            + ",".join(
                _encode_value(item, f"{at}[{index}]", depth + 1)
                for index, item in enumerate(value)
            )
            + "]"
        )
    if isinstance(value, dict):
        for key in value:
            if not isinstance(key, str):
                raise _DigestInputError(
                    "digest_input_unsupported", f"{at}: map key {key!r} is not a string"
                )
        keys = sorted(value, key=_utf16_key)
        return (
            "{"
            + ",".join(
                _encode_string(key, f"{at} key")
                + ":"
                + _encode_value(value[key], f"{at}.{key}", depth + 1)
                for key in keys
            )
            + "}"
        )
    raise _DigestInputError(
        "digest_input_unsupported",
        f"{at}: {type(value).__name__} is outside the digest value model",
    )


def _encode_number(value: Union[int, float], at: str) -> str:
    if isinstance(value, float) and (
        not math.isfinite(value) or not value.is_integer()
    ):
        raise _DigestInputError(
            "digest_input_float",
            f"{at}: {value!r} is not an integer; render it as a string if it belongs in a digest",
        )
    if abs(value) > _MAX_SAFE_INTEGER:
        raise _DigestInputError(
            "digest_input_int_range",
            f"{at}: {describe_number(value)} is outside ±(2^53 - 1)",
        )
    # int() also writes a negative zero float as "0".
    return str(int(value))


def _encode_string(value: str, at: str) -> str:
    out = ['"']
    for index, char in enumerate(value):
        point = ord(char)
        if 0xD800 <= point <= 0xDFFF:
            raise _DigestInputError(
                "digest_input_unsupported",
                f"{at}: string has an unpaired surrogate at index {index}",
            )
        escaped = _SHORT_ESCAPES.get(point)
        if escaped is not None:
            out.append(escaped)
        elif point < 0x20:
            out.append("\\u%04x" % point)
        else:
            out.append(char)
    out.append('"')
    return "".join(out)


# -------------------------------------------------------------- reuse verdict

# The status vocabulary every artifact is read through. Only "complete" is
# reusable.
ArtifactStatus = Literal["complete", "partial", "failed", "blocked"]

# Used when the caller passes no status_map: the vocabulary maps to itself.
_IDENTITY_STATUS_MAP: Mapping[str, ArtifactStatus] = {
    "complete": "complete",
    "partial": "partial",
    "failed": "failed",
    "blocked": "blocked",
}


def _text(value: Any, what: str) -> Optional[str]:
    """An optional string argument: None when absent, refused when not text."""
    if value is None or isinstance(value, str):
        return value
    raise TypeError(f"{what} must be a string or null, got {type(value).__name__}")


def _is_map_of(value: Any, allowed: Any) -> bool:
    return isinstance(value, dict) and all(allowed(item) for item in value.values())


def evaluate_checkpoint_artifact(
    *,
    stage_id: str,
    artifact: Optional[Mapping[str, Any]] = None,
    subject_ref: Optional[str] = None,
    expected_contract_revision: Optional[str] = None,
    expected_stage_config_digest: Optional[str] = None,
    expected_dependency_digests: Optional[Mapping[str, str]] = None,
    required_resume_from_stage: Optional[str] = None,
    validation_issues: Optional[list[Mapping[str, Any]]] = None,
    status_map: Optional[Mapping[str, ArtifactStatus]] = None,
) -> list[dict[str, Any]]:
    """Decide whether a recorded artifact may be reused.

    Returns every issue found, in this order: status, contract revision,
    stage-config digest, dependency digests (by UTF-16 key order), the
    caller's validation issues. With none, returns one "valid" issue, so the
    result is never empty. `artifact` is None when it does not exist;
    `status_map` replaces the default identity map, and a status it does not
    name is not reusable. An expectation that is None is not checked; any
    string is compared, the empty one included.
    """
    if not isinstance(stage_id, str):
        raise TypeError(f"stage_id must be a string, got {type(stage_id).__name__}")
    subject_ref = _text(subject_ref, "subject_ref")
    resume = _text(required_resume_from_stage, "required_resume_from_stage")
    if resume is None:
        resume = stage_id
    expected_contract_revision = _text(
        expected_contract_revision, "expected_contract_revision"
    )
    expected_stage_config_digest = _text(
        expected_stage_config_digest, "expected_stage_config_digest"
    )
    if artifact is not None and not isinstance(artifact, dict):
        raise TypeError("artifact must be a map or null")
    if expected_dependency_digests is None:
        expected_dependency_digests = {}
    elif not _is_map_of(expected_dependency_digests, lambda v: isinstance(v, str)):
        raise TypeError("expected_dependency_digests must map ids to strings")
    if status_map is None:
        status_map = _IDENTITY_STATUS_MAP
    elif not _is_map_of(
        status_map, lambda v: isinstance(v, str) and v in _IDENTITY_STATUS_MAP
    ):
        raise TypeError("status_map must map statuses to the vocabulary")
    if validation_issues is None:
        validation_issues = []
    elif not isinstance(validation_issues, list) or not all(
        isinstance(issue, dict) for issue in validation_issues
    ):
        raise TypeError("validation_issues must be a list of maps")

    def base(
        level: VerdictLevel, reason: str, message: str, **details: Any
    ) -> dict[str, Any]:
        return {
            **verdict(level, reason, f"{stage_id}: {message}", resume),
            "stage_id": stage_id,
            "subject_ref": subject_ref,
            **details,
        }

    # A caller's issue is an argument like the rest: what is outside the contract is refused before
    # anything is judged, an absent artifact included. A null field is an absent one.
    for validation_issue in validation_issues:
        for key, field in validation_issue.items():
            if field is not None:
                _holds_what_a_verdict_promises(key, field)

    if artifact is None:
        return [base("halt", "artifact_missing", "nothing was recorded")]

    issues: list[dict[str, Any]] = []
    status = artifact.get("status")
    if _js_falsy(status):
        issues.append(
            base("halt", "artifact_status_missing", "what was recorded has no status")
        )
    elif _resolve_status(status, status_map) != "complete":
        issues.append(
            base(
                "halt",
                "artifact_status_not_reusable",
                "the recorded status is not one that may be reused",
                actual_status=status,
            )
        )

    revision = artifact.get("contract_revision")
    if expected_contract_revision is not None and _js_falsy(revision):
        issues.append(
            base(
                "halt",
                "contract_revision_missing",
                f"a contract revision of {expected_contract_revision!r} is expected"
                " and none is recorded",
                expected_contract_revision=expected_contract_revision,
            )
        )
    elif expected_contract_revision is not None and not _same_text(
        revision, expected_contract_revision
    ):
        issues.append(
            base(
                "halt",
                "contract_revision_mismatch",
                "the recorded contract revision is not the expected"
                f" {expected_contract_revision!r}",
                expected_contract_revision=expected_contract_revision,
                actual_contract_revision=revision,
            )
        )

    config = artifact.get("stage_config_digest")
    if expected_stage_config_digest is not None and not _same_text(
        config, expected_stage_config_digest
    ):
        issues.append(
            base(
                "halt",
                "stage_config_digest_mismatch",
                "the recorded stage config digest is not the expected"
                f" {expected_stage_config_digest!r}",
                expected_stage_config_digest=expected_stage_config_digest,
                actual_stage_config_digest=config,
            )
        )

    expected_dependencies = expected_dependency_digests
    recorded = artifact.get("dependency_digests")
    for dependency_id in sorted(expected_dependencies, key=_utf16_key):
        expected_digest = expected_dependencies[dependency_id]
        actual_digest = (
            recorded.get(dependency_id) if isinstance(recorded, dict) else None
        )
        if not _same_text(actual_digest, expected_digest):
            issues.append(
                base(
                    "halt",
                    "dependency_digest_mismatch",
                    f"the dependency {dependency_id!r} moved",
                    dependency_id=dependency_id,
                    expected_digest=expected_digest,
                    actual_digest=actual_digest,
                )
            )

    for validation_issue in validation_issues:
        given = {
            key: field for key, field in validation_issue.items() if field is not None
        }
        issues.append(
            {
                **base(
                    "halt",
                    "validation_issue",
                    "the caller's own validation found something",
                ),
                **given,
            }
        )

    if not issues:
        issues.append(base("ok", "checkpoint_valid", "the artifact may be reused"))
    # An ok verdict has no resume, whoever set it.
    for issue in issues:
        if issue["verdict"] == "ok":
            issue["resume"] = None
    return issues


_TEXT_FIELDS = ("reason", "message", "resume", "stage_id", "subject_ref")


def _holds_what_a_verdict_promises(key: str, field: Any) -> None:
    """Refuse a caller's field that would leave the issue outside the shape it
    is laid over. Everything the verdict does not name is the caller's own and
    is not judged."""
    if key == "spec":
        # A caller may disagree with a verdict, not sign one.
        raise TypeError(
            "a validation issue carries a spec, which only the library says"
        )
    if key == "verdict" and not is_verdict_level(field):
        raise TypeError("a validation issue's verdict is not one of the three")
    if key in _TEXT_FIELDS and not isinstance(field, str):
        raise TypeError(f"a validation issue's {key} is not text")


def _resolve_status(
    status: Any, status_map: Mapping[str, ArtifactStatus]
) -> Optional[str]:
    if not isinstance(status, str):
        return None
    return status_map.get(status)


def _js_falsy(value: Any) -> bool:
    """JavaScript falsiness over JSON values: null, false, 0, NaN and "".

    An empty list or dict is truthy in JavaScript, unlike Python.
    """
    if isinstance(value, (int, float)):  # a bool is an int: False is 0
        return value == 0 or value != value
    return value is None or value == ""


def _same_text(recorded: Any, expected: str) -> bool:
    """What an artifact records equals an expectation only when it is the same
    string: a recorded 1 is not "1", and a recorded ["v2"] is not "v2"."""
    return isinstance(recorded, str) and recorded == expected

File: rust/haltrule/src/checkpoint.rs
//! Checkpoint - canonical form and digest of a value, and a verdict on whether
//! an artifact an earlier run recorded may be reused.
//!
//! The arguments are typed, and an argument of another type cannot be built:
//! `None` is absent, and nothing else is. What the artifact records is data,
//! written by whoever wrote it, and is read leniently instead.

use alloc::borrow::ToOwned;
use alloc::collections::BTreeMap;
use alloc::format;
use alloc::string::String;
use alloc::vec;
use alloc::vec::Vec;
use core::cmp::Ordering;

use sha2::{Digest, Sha256};

use crate::value::{Map, Refused, Value};
use crate::verdict::{verdict, Level, Verdict, SPEC};

/// The largest integer every implementation represents exactly: 2^53 - 1.
const MAX_SAFE_INTEGER: i64 = (1 << 53) - 1;

/// The same bound as a double. It is exact: 2^53 - 1 is representable.
const MAX_SAFE_INTEGER_AS_FLOAT: f64 = MAX_SAFE_INTEGER as f64;

/// The deepest nesting of lists and maps a digest input may have. A bound
/// every language reaches without exhausting its stack, so all of them halt at
/// the same depth instead of each crashing at its own.
const MAX_DEPTH: usize = 100;

const DIGEST_INPUT_FLOAT: &str = "digest_input_float";
const DIGEST_INPUT_INT_RANGE: &str = "digest_input_int_range";
const DIGEST_INPUT_UNSUPPORTED: &str = "digest_input_unsupported";

/// A value outside the digest value model: a halt verdict whose message names
/// where the value sits, a path being no thing two languages spell alike.
fn halted(reason: &'static str, message: String) -> Verdict {
    verdict(Level::Halt, reason, message)
}

/// The canonical form of a value: an RFC 8785 subset. Map keys are sorted by
/// UTF-16 code unit, there is no insignificant whitespace, integers are
/// written in decimal, and strings carry the spec's escapes. Strings are not
/// Unicode-normalized.
pub fn canonicalize(value: &Value) -> Result<String, Verdict> {
    encode_value(value, "$", 0)
}

/// "sha256:" and the lowercase hex SHA-256 of the canonical form's UTF-8
/// bytes, or the halt [`canonicalize`] returned.
pub fn checkpoint_digest(value: &Value) -> Result<String, Verdict> {
    let canonical = canonicalize(value)?;
    let sum = Sha256::digest(canonical.as_bytes());
    let mut out = String::from("sha256:");
    for byte in sum {
        out.push(hex_digit(byte >> 4));
        out.push(hex_digit(byte & 0x0f));
    }
    Ok(out)
}

fn hex_digit(nibble: u8) -> char {
    char::from(b"0123456789abcdef"[nibble as usize])
}

fn encode_value(value: &Value, at: &str, depth: usize) -> Result<String, Verdict> {
    // The match is exhaustive because `Value` is closed, so there is no other
    // kind to meet: what the reference answers for below its last case cannot
    // be handed to this port at all.
    match value {
        Value::Null => Ok("null".to_owned()),
        Value::Bool(held) => Ok(if *held { "true" } else { "false" }.to_owned()),
        Value::Int(held) => {
            if *held > MAX_SAFE_INTEGER || *held < -MAX_SAFE_INTEGER {
                return Err(halted(
                    DIGEST_INPUT_INT_RANGE,
                    format!("{at}: {held} is outside +/-(2^53 - 1); render it as a string if it belongs in a digest"),
                ));
            }
            Ok(format!("{held}"))
        }
        Value::Float(held) => encode_float(*held, at),
        Value::Text(held) => Ok(encode_string(held)),
        Value::List(held) => {
            if depth >= MAX_DEPTH {
                return Err(too_deep(at));
            }
            let mut out = String::from("[");
            for (index, item) in held.iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                out.push_str(&encode_value(item, &format!("{at}[{index}]"), depth + 1)?);
            }
            out.push(']');
            Ok(out)
        }
        Value::Map(held) => {
            if depth >= MAX_DEPTH {
                return Err(too_deep(at));
            }
            let mut out = String::from("{");
            for (index, (key, member)) in sorted_entries(held).into_iter().enumerate() {
                if index > 0 {
                    out.push(',');
                }
                out.push_str(&encode_string(key));
                out.push(':');
                out.push_str(&encode_value(member, &format!("{at}.{key}"), depth + 1)?);
            }
            out.push('}');
            Ok(out)
        }
    }
}

fn too_deep(at: &str) -> Verdict {
    halted(
        DIGEST_INPUT_UNSUPPORTED,
        format!("{at}: nested deeper than {MAX_DEPTH} lists and maps"),
    )
}

/// A number the caller held as a double. The reference asks whether it is an
/// integer before asking whether it is in range; this asks the range first,
/// and meets the same halt on every input. Above 2^52 the doubles are spaced a
/// whole apart, so every double whose magnitude passes 2^53 - 1 is an integer
/// already and could only ever have failed the range test. Ruling the range
/// out first is also what makes the integrality test below exact without
/// `std`: `trunc` and `floor` are the standard library's, but within the range
/// a cast to `i64` truncates and casts back losing nothing, so a number that
/// survives the round trip unchanged is the integer it looks like. A negative
/// zero survives it as zero, which is the answer wanted.
fn encode_float(number: f64, at: &str) -> Result<String, Verdict> {
    if number.is_nan() || number.is_infinite() {
        return Err(not_an_integer(number, at));
    }
    if number > MAX_SAFE_INTEGER_AS_FLOAT || number < -MAX_SAFE_INTEGER_AS_FLOAT {
        return Err(halted(
            DIGEST_INPUT_INT_RANGE,
            format!("{at}: {number} is outside +/-(2^53 - 1); render it as a string if it belongs in a digest"),
        ));
    }
    let whole = number as i64;
    if number != whole as f64 {
        return Err(not_an_integer(number, at));
    }
    // A negative zero is written "0", as every zero is.
    Ok(format!("{whole}"))
}

fn not_an_integer(number: f64, at: &str) -> Verdict {
    halted(
        DIGEST_INPUT_FLOAT,
        format!(
            "{at}: {number} is not an integer; render it as a string if it belongs in a digest"
        ),
    )
}

/// A map's members in the spec's order, which is by UTF-16 code unit and not
/// Rust's own, because they disagree above U+FFFF.
fn sorted_entries(held: &Map) -> Vec<(&String, &Value)> {
    let mut entries: Vec<(&String, &Value)> = held.iter().collect();
    // The entries arrive in one order already - a `BTreeMap`'s, by UTF-8 byte -
    // and the sort below is stable, so they are put in one order before being
    // put in the spec's. The answer is the same either way; what this settles
    // is the work done to reach it.
    entries.sort_by(|left, right| cmp_utf16(left.0, right.0));
    entries
}

fn cmp_utf16(left: &str, right: &str) -> Ordering {
    let mut left_units = left.encode_utf16();
    let mut right_units = right.encode_utf16();
    loop {
        match (left_units.next(), right_units.next()) {
            (Some(left_unit), Some(right_unit)) => {
                if left_unit != right_unit {
                    return left_unit.cmp(&right_unit);
                }
            }
            (None, None) => return Ordering::Equal,
            (None, Some(_)) => return Ordering::Less,
            (Some(_), None) => return Ordering::Greater,
        }
    }
}

/// Writes a string of Unicode scalar values, quoted, with exactly seven
/// two-character escapes, every other code point below U+0020 as `\u00xx` in
/// lowercase hex, and everything else as itself.
///
/// The reference halts on a string that is not made of scalar values, because
/// an unpaired surrogate has no UTF-8 spelling. A Rust `str` is made of scalar
/// values by construction, so that halt has nothing to catch here and this
/// cannot fail: the case is one the port's types shut out before it ran.
fn encode_string(value: &str) -> String {
    let mut out = String::from("\"");
    for character in value.chars() {
        match character {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{08}' => out.push_str("\\b"),
            '\u{0c}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            _ => {
                if character < '\u{20}' {
                    out.push_str(&format!("\\u{:04x}", character as u32));
                } else {
                    out.push(character);
                }
            }
        }
    }
    out.push('"');
    out
}

// ------------------------------------------------------------ reuse verdict

/// The vocabulary every artifact is read through. Only [`ArtifactStatus::Complete`]
/// is reusable.
///
/// The reference holds a status as a string and refuses a `status_map` that
/// answers outside the vocabulary. This enum is closed, so an answer outside
/// it cannot be built: the same contract, kept by the type instead of by code.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ArtifactStatus {
    Complete,
    Partial,
    Failed,
    Blocked,
}

impl ArtifactStatus {
    /// The default map, which sends each word of the vocabulary to itself, and
    /// the only way to name a status from text: a word outside the vocabulary
    /// is `None`, which is where a caller's bad `status_map` value is refused.
    pub fn from_name(name: &str) -> Option<Self> {
        match name {
            "complete" => Some(Self::Complete),
            "partial" => Some(Self::Partial),
            "failed" => Some(Self::Failed),
            "blocked" => Some(Self::Blocked),
            _ => None,
        }
    }
}

/// One verdict about an artifact: the five fields of a [`Verdict`], with
/// `stage_id` and `subject_ref` beside them, and what its reason adds. A
/// caller's own issue may lay any of them over the defaults, so it is a map
/// and not the struct, and its `reason` is text and not a `&'static str`.
pub type Issue = Map;

/// What the caller expects of an artifact now. `None` is absent; an empty map
/// is present.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct CheckpointArgs {
    pub stage_id: String,
    pub subject_ref: Option<String>,
    /// What was recorded; `None` when nothing was.
    pub artifact: Option<Map>,
    /// An absent expectation is not checked; any string is compared, the empty
    /// one included.
    pub expected_contract_revision: Option<String>,
    pub expected_stage_config_digest: Option<String>,
    pub expected_dependency_digests: Option<BTreeMap<String, String>>,
    /// Defaults to `stage_id`.
    pub required_resume_from_stage: Option<String>,
    pub validation_issues: Option<Vec<Map>>,
    /// Replaces the default map, which sends each word of the vocabulary to
    /// itself; a status it does not name is not reusable.
    pub status_map: Option<BTreeMap<String, ArtifactStatus>>,
}

fn text(value: Option<&str>) -> Value {
    match value {
        None => Value::Null,
        Some(held) => Value::Text(held.to_owned()),
    }
}

/// Decides whether a recorded artifact may be reused. It answers every issue
/// found, in this order: status, contract revision, stage-config digest,
/// dependency digests by UTF-16 key order, the caller's validation issues.
/// With none it answers one valid issue, so the answer is never empty.
///
/// The reference also refuses a status map that answers outside the
/// vocabulary. Here [`ArtifactStatus`] is closed and such a map cannot be
/// built, so that refusal has nothing to catch; what is left to refuse is a
/// caller's own validation issue that signs the spec or gives a verdict
/// outside the three.
pub fn evaluate_checkpoint_artifact(args: &CheckpointArgs) -> Result<Vec<Issue>, Refused> {
    let resume = args
        .required_resume_from_stage
        .as_deref()
        .unwrap_or(&args.stage_id);
    let base = |level: Level, reason: &str, message: String| -> Issue {
        let mut issue = Issue::new();
        issue.insert("spec".to_owned(), Value::Text(SPEC.to_owned()));
        issue.insert("verdict".to_owned(), Value::Text(level.as_str().to_owned()));
        issue.insert("reason".to_owned(), Value::Text(reason.to_owned()));
        issue.insert(
            "message".to_owned(),
            Value::Text(format!("{}: {}", args.stage_id, message)),
        );
        issue.insert("resume".to_owned(), Value::Text(resume.to_owned()));
        issue.insert("stage_id".to_owned(), Value::Text(args.stage_id.clone()));
        issue.insert("subject_ref".to_owned(), text(args.subject_ref.as_deref()));
        issue
    };

    // A caller's issue is an argument like the rest: what is outside the contract is refused before
    // anything is judged, an absent artifact included. A null field is an absent one.
    for given in args.validation_issues.iter().flatten() {
        for (key, field) in given {
            if *field != Value::Null {
                holds_what_a_verdict_promises(key, field)?;
            }
        }
    }

    let artifact = match args.artifact.as_ref() {
        None => {
            return Ok(vec![base(
                Level::Halt,
                "artifact_missing",
                "nothing was recorded".to_owned(),
            )])
        }
        Some(held) => held,
    };

    let mut issues: Vec<Issue> = Vec::new();

    let status = artifact.get("status");
    if js_falsy(status) {
        issues.push(base(
            Level::Halt,
            "artifact_status_missing",
            "what was recorded has no status".to_owned(),
        ));
    } else if resolve_status(status, args.status_map.as_ref()) != Some(ArtifactStatus::Complete) {
        let mut issue = base(
            Level::Halt,
            "artifact_status_not_reusable",
            "the recorded status is not one that may be reused".to_owned(),
        );
        issue.insert("actual_status".to_owned(), recorded_value(status));
        issues.push(issue);
    }

    let revision = artifact.get("contract_revision");
    if let Some(expected) = args.expected_contract_revision.as_deref() {
        if js_falsy(revision) {
            let mut issue = base(
                Level::Halt,
                "contract_revision_missing",
                format!("a contract revision of {expected:?} is expected and none is recorded"),
            );
            issue.insert(
                "expected_contract_revision".to_owned(),
                Value::Text(expected.to_owned()),
            );
            issues.push(issue);
        } else if !same_text(revision, expected) {
            let mut issue = base(
                Level::Halt,
                "contract_revision_mismatch",
                format!("the recorded contract revision is not the expected {expected:?}"),
            );
            issue.insert(
                "expected_contract_revision".to_owned(),
                Value::Text(expected.to_owned()),
            );
            issue.insert(
                "actual_contract_revision".to_owned(),
                recorded_value(revision),
            );
            issues.push(issue);
        }
    }

    let config = artifact.get("stage_config_digest");
    if let Some(expected) = args.expected_stage_config_digest.as_deref() {
        if !same_text(config, expected) {
            let mut issue = base(
                Level::Halt,
                "stage_config_digest_mismatch",
                format!("the recorded stage config digest is not the expected {expected:?}"),
            );
            issue.insert(
                "expected_stage_config_digest".to_owned(),
                Value::Text(expected.to_owned()),
            );
            issue.insert(
                "actual_stage_config_digest".to_owned(),
                recorded_value(config),
            );
            issues.push(issue);
        }
    }

    // A recorded `dependency_digests` that is not a map records nothing.
    let recorded = match artifact.get("dependency_digests") {
        Some(Value::Map(held)) => Some(held),
        _ => None,
    };
    if let Some(expected_digests) = args.expected_dependency_digests.as_ref() {
        let mut expected_ids: Vec<(&String, &String)> = expected_digests.iter().collect();
        expected_ids.sort_by(|left, right| cmp_utf16(left.0, right.0));
        for (id, expected) in expected_ids {
            let actual = recorded.and_then(|held| held.get(id.as_str()));
            if !same_text(actual, expected) {
                let mut issue = base(
                    Level::Halt,
                    "dependency_digest_mismatch",
                    format!("the dependency {id:?} moved"),
                );
                issue.insert("dependency_id".to_owned(), Value::Text(id.clone()));
                issue.insert("expected_digest".to_owned(), Value::Text(expected.clone()));
                issue.insert("actual_digest".to_owned(), recorded_value(actual));
                issues.push(issue);
            }
        }
    }

    for given in args.validation_issues.iter().flatten() {
        let mut issue = base(
            Level::Halt,
            "validation_issue",
            "the caller's own validation found something".to_owned(),
        );
        for (key, field) in given {
            if *field == Value::Null {
                continue;
            }
            issue.insert(key.clone(), field.clone());
        }
        issues.push(issue);
    }

    if issues.is_empty() {
        issues.push(base(
            Level::Ok,
            "checkpoint_valid",
            "the artifact may be reused".to_owned(),
        ));
    }
    // An ok verdict has no resume, whoever set it.
    for issue in issues.iter_mut() {
        if issue.get("verdict") == Some(&Value::Text(Level::Ok.as_str().to_owned())) {
            issue.insert("resume".to_owned(), Value::Null);
        }
    }
    Ok(issues)
}

/// Refuses a caller's field that would leave the issue outside the shape it is
/// laid over. Everything the verdict does not name is the caller's own and is
/// not judged.
fn holds_what_a_verdict_promises(key: &str, field: &Value) -> Result<(), Refused> {
    match key {
        // A caller may disagree with a verdict, not sign one.
        "spec" => Err(Refused(
            "a validation issue carries a spec, which only the library says",
        )),
        "verdict" if !matches!(field, Value::Text(held) if known_level(held)) => Err(Refused(
            "a validation issue's verdict is not one of the three",
        )),
        "reason" | "message" | "resume" | "stage_id" | "subject_ref"
            if !matches!(field, Value::Text(_)) =>
        {
            Err(Refused("a validation issue's field is not text"))
        }
        _ => Ok(()),
    }
}

/// Whether a level is one of the three. The library never asks it of itself;
/// a caller's own verdict, laid over a checkpoint issue, is the one place a
/// level arrives from outside, and it arrives as text.
fn known_level(named: &str) -> bool {
    named == Level::Ok.as_str() || named == Level::Warning.as_str() || named == Level::Halt.as_str()
}

fn resolve_status(
    status: Option<&Value>,
    status_map: Option<&BTreeMap<String, ArtifactStatus>>,
) -> Option<ArtifactStatus> {
    let named = match status {
        Some(Value::Text(held)) => held,
        _ => return None,
    };
    match status_map {
        None => ArtifactStatus::from_name(named),
        Some(held) => held.get(named.as_str()).copied(),
    }
}

/// What a recorded value is, for an `actual_` member: what the artifact holds,
/// null when it holds nothing.
fn recorded_value(value: Option<&Value>) -> Value {
    value.cloned().unwrap_or(Value::Null)
}

/// What "absent" means of a recorded value: null, false, 0, NaN and the empty
/// string. An empty list or map is present. A field the artifact does not
/// record is null, which is absent as everywhere else.
fn js_falsy(value: Option<&Value>) -> bool {
    match value {
        None | Some(Value::Null) => true,
        Some(Value::Bool(held)) => !*held,
        Some(Value::Int(held)) => *held == 0,
        Some(Value::Float(held)) => *held == 0.0 || held.is_nan(),
        Some(Value::Text(held)) => held.is_empty(),
        _ => false,
    }
}

/// What equality means between a recorded value and an expectation: the same
/// string, so a recorded 1 is not "1" and a recorded ["v2"] is not "v2".
fn same_text(recorded: Option<&Value>, expected: &str) -> bool {
    matches!(recorded, Some(Value::Text(held)) if held == expected)
}
